File: src/extras-kio/libjob.cpp
#include <QByteArray>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QUrl>
#include <job.h>
#include "libjob.h"
#include "libjob.hxx"

libqt_string KIO_BuildErrorString(int param1, const libqt_string param2) {
    QString param2_QString = QString::fromUtf8(param2.data, param2.len);
    QString _ret = KIO::buildErrorString(static_cast<int>(param1), param2_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KIO_RawErrorDetail(int param1, const libqt_string param2, const QUrl* param3, int param4) {
    QString param2_QString = QString::fromUtf8(param2.data, param2.len);
    QByteArray _qb = KIO::rawErrorDetail(static_cast<int>(param1), param2_QString, param3, static_cast<int>(param4));
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _qb.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

File: src/extras-kio/libjob.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/kio.html
pub const kio = struct {
    /// [Qt documentation](https://api.kde.org/kio.html#buildErrorString)
    ///
    /// ``` param1: i32, param2: []const u8, allocator: std.mem.Allocator ```
    pub fn BuildErrorString(param1: i32, param2: []const u8, allocator: std.mem.Allocator) []const u8 {
        const param2_str = qtc.libqt_string{
            .len = param2.len,
            .data = param2.ptr,
        };
        const _str = qtc.KIO_BuildErrorString(@intCast(param1), param2_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kio.BuildErrorString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/kio.html#rawErrorDetail)
    ///
    /// ``` param1: i32, param2: []const u8, param3: QtC.QUrl, param4: i32, allocator: std.mem.Allocator ```
    pub fn RawErrorDetail(param1: i32, param2: []const u8, param3: ?*anyopaque, param4: i32, allocator: std.mem.Allocator) []u8 {
        const param2_str = qtc.libqt_string{
            .len = param2.len,
            .data = param2.ptr,
        };
        const _bytearray: qtc.libqt_string = qtc.KIO_RawErrorDetail(@intCast(param1), param2_str, @ptrCast(param3), @intCast(param4));
        defer qtc.libqt_string_free(&_bytearray);
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("kio.RawErrorDetail: Memory allocation failed");
        @memcpy(_ret, _bytearray.data[0.._bytearray.len]);
        return _ret;
    }
};
